### PR TITLE
Add handling to user + meeting id checks to support activity restores

### PR DESCRIPTION
**Pull request by keevan**
- This may happen when the activity is restored into a new course, but the recording is very much tied to the previous course's information. This loops through (based on meeting id) and checks if the user has permissions to view this recording.

### CODE REVIEW

File: proxy_presentation.php [hunk 1]
```diff
@@ -60,14 +60,37 @@
...
-    require_login($viewinstance['course'], true, $viewinstance['cm']);
+
+    // Check if the user has appropriate access to the meeting resources.
+    $meetingsinstances = $DB->get_records('bigbluebuttonbn', ['meetingid' => $basemeetingid], '', 'id, course');
+    $access = false;
+    $counter = 1;
+    $maxcount = count($meetingsinstances);
+    foreach ($meetingsinstances as $bbbinstance) {
+        try {
+            $viewinstance = bigbluebuttonbn_view_instance_bigbluebuttonbn($bbbinstance->id);
```
Comment (Peterburnett) [inline proxy_presentation.php:78]: Would it be more performant to pull just the required metadata straight from the DB?

File: proxy_presentation.php [hunk 2]
```diff
@@ -60,14 +60,37 @@
...
+            // NOTE: Bit dirty, but felt better than pulling out specific internals.
+            // And does a normal redirect on the last check (e.g. if it fails).
+            require_login($viewinstance['course'], true, $viewinstance['cm'], $setwantsurltome = true, $preventredirect = ($counter !== $maxcount));
```
Comment (Peterburnett) [inline proxy_presentation.php:81]: I think we should use has_capability here instead, and just have a boolean check, and we can just break on the first true response, else exception

Comment (Peterburnett) [reply proxy_presentation.php:81]: and then require_login after the break out of the loop once we find a matching activity

File: proxy_presentation.php (hunk 1 above)
Comment (keevan) [reply proxy_presentation.php:85]: That BBB helper function does just that (albeit in several small queries)
```
function bigbluebuttonbn_view_instance_bigbluebuttonbn($bigbluebuttonbnid) {
    global $DB;
    $bigbluebuttonbn = $DB->get_record('bigbluebuttonbn', array('id' => $bigbluebuttonbnid), '*', MUST_EXIST);
    $course = $DB->get_record('course', array('id' => $bigbluebuttonbn->course), '*', MUST_EXIST);
    $cm = get_coursemodule_from_instance('bigbluebuttonbn', $bigbluebuttonbn->id, $course->id, false, MUST_EXIST);
    return array('cm' => $cm, 'course' => $course, 'bigbluebuttonbn' => $bigbluebuttonbn);
}
```

File: proxy_presentation.php (hunk 2 above)
Comment (keevan) [reply proxy_presentation.php:86]: Done, except kept the loop, only attempts require_login on initial pass and breaks loops on exceptions or if the require_login passes

Comment (keevan) [reply proxy_presentation.php]: Removed loop, was less lines and clearer intent without. No need to explicitly throw exception, require_login does that for us. On capability checks, as-is there will be more DB lookups if require_login checks was performed as is. Did you want me to explicitly throw an error/exception regarding access if the user does not have capabilities for any of the returning activities?

File: proxy_presentation.php [hunk 3]
```diff
@@ -60,14 +60,36 @@
...
+        // Checks the BBB instances to see if the user has capabilities for viewing - breaking on first success.
+        if (bigbluebuttonbn_has_capability($bbbinstance->id, 'mod/bigbluebuttonbn:view')) {
+            $capable = true;
+            break;
+        }
+    }
+
+    if (!$capable) {
+        if (!$cm = get_coursemodule_from_instance('bigbluebuttonbn', $bbbinstance->id)) {
+            return false;
```
Comment (Peterburnett) [inline proxy_presentation.php:84]: I think this should just exception as well same as the line below

**Pull request merged**